test/test_device_batch_copy.cu

```diff
@@ -0,0 +1,427 @@
+                                        stream));
+
+  // Record event after algorithm
+  cudaEventRecord(events[1], stream);
+
+  // Copy back the output buffer
+  CubDebugExit(
+    cudaMemcpyAsync(h_gpu_results.get(), d_out, num_total_bytes, cudaMemcpyDeviceToHost, stream));
+
+  // Make sure results have been copied back to the host
+  CubDebugExit(cudaStreamSynchronize(stream));
+
+  // CPU-side result generation for verification
+  for (BufferOffsetT i = 0; i < num_buffers; i++)
+  {
+    std::copy(d_range_srcs[i],
+              d_range_srcs[i] + h_buffer_sizes[i],
+              h_out.get() + h_buffer_dst_offsets[i]);
+  }
+
+  float duration = 0;
+  cudaEventElapsedTime(&duration, events[0], events[1]);
+
+#ifdef CUB_TEST_BENCHMARK
+  size_t stats_src_offsets = sizeof(ByteOffsetT) * num_buffers;
+  size_t stats_dst_offsets = sizeof(ByteOffsetT) * num_buffers;
+  size_t stats_sizes       = sizeof(BufferSizeT) * num_buffers;
+  size_t stats_data_copied = 2 * num_total_bytes;
+
+  std::cout
+    << "Min. buffer size: " << min_buffer_size << ", max. buffer size: " << max_buffer_size     //
+    << ", num_buffers: " << num_buffers                                                         //
+    << ", out_gen: " << ((output_gen == TestDataGen::RANDOM) ? "SHFL" : "CONSECUTIVE")          //
+    << ", src size: " << stats_src_offsets << ", dst size: " << stats_dst_offsets               //
+    << ", sizes size: " << stats_sizes << ", cpy_data_size: " << stats_data_copied              //
+    << ", total: " << (stats_src_offsets + stats_dst_offsets + stats_sizes + stats_data_copied) //
+    << ", duration: " << duration                                                               //
+    << ", BW: "
+    << ((double)(stats_src_offsets + stats_dst_offsets + stats_sizes + stats_data_copied) /
+        1000000000.0) /
+         (duration / 1000.0)
+    << "GB/s \n";
+#endif
+
+  const auto it_pair =
+    std::mismatch(h_gpu_results.get(), h_gpu_results.get() + num_total_items, h_out.get());
+
+  if (it_pair.first != h_gpu_results.get() + num_total_items)
+  {
+    std::cout << "Mismatch at index " << it_pair.first - h_gpu_results.get()
+              << ", CPU vs. GPU: " << *it_pair.second << ", " << *it_pair.first << "\n";
+  }
+  AssertEquals(it_pair.first, h_gpu_results.get() + num_total_items);
+
+  CubDebugExit(cudaFree(d_out));
+  CubDebugExit(cudaFree(d_buffer_dst_offsets));
+  CubDebugExit(cudaFree(d_buffer_sizes));
+  CubDebugExit(cudaFree(d_temp_storage));
+}
+
+int main(int argc, char **argv)
+{
+  CommandLineArgs args(argc, argv);
+
+  // Initialize device
+  CubDebugExit(args.DeviceInit());
+
+  //---------------------------------------------------------------------
+  // DeviceCopy::Batched tests
+  //---------------------------------------------------------------------
+  // The most granular type being copied. Buffer's will be aligned and their size be an integer
+  // multiple of this type
+  using AtomicCopyT = int64_t;
```

Comment:
Since this is even more generic than the `DeviceMemcpy::Batched` it would be great if we could test for different item types. Like, `char`, `int64_t`, and one non-built-in type. For the non-built-in type you could use a `thrust::zip_iterator` to give you `thrust::tuple`s - or a different custom type, whichever you prefer.

Reply:
The test code currently includes different types such as int64_t, tuple<int64_t, int32_t int16_t, char, char>, uint8_t (at the end for the big buffer test),. The current code uses fancy iterators for both input and output.